### PR TITLE
Compression (DBTL-1)

### PR DESCRIPTION
**Pull request by seb-hyland**
**This code is heavily WIP at the moment and is untested. Exec.rs likely contains bugs or problems.**

The goal of this branch is to create:

1. A function that can arbitrarily run binaries on Windows/Linux (the binary will be written into a variable at compile time, then ran as an anonymous file at runtime)
2. Create the DBTL-1 variant of the compression implementation.

### CODE REVIEW

**Comment by seb-hyland** [comment]
Provides compression and decompression mechanisms for lz4 and ts_zip. Exec.rs has been removed.

Compression takes `inpath:PathBuf` as input and outputs `Result<PathBuf, io::Error>` where `PathBuf` references the output file.
Decompression takes `inpath:PathBuf` and `outpath:PathBuf` as input and outputs `Result<()>`.

File: src-tauri/src/compressor.rs [hunk 1]
```diff
@@ -1,12 +1,141 @@
...
+        );
+
+        let input_file = File::open(inpath.as_path())?;
+        let mut decoder = Decoder::new(input_file)?;
+        let mut output_file = File::create(outpath.as_path())?;
+        io::copy(&mut decoder, &mut output_file)?;
+
+        Ok(())
     }
 }
```
Comment (lhao03) [inline src-tauri/src/compressor.rs:141]: can you add a test with quickcheck, just to quickly verify the decompress and compress work? 

Example: https://github.com/UBC-iGEM/dna-software/blob/main/src-tauri/src/encoder.rs#L75-L93

**Review by lhao03 — CHANGES_REQUESTED** [review]

File: src-tauri/src/compressor.rs [hunk 2]
```diff
@@ -1,12 +1,141 @@
...
+use lz4::{Decoder, EncoderBuilder};
+use std::{
+    fs::File,
+    io::{self, BufRead, BufReader, Result, Write},
+    path::{Path, PathBuf},
+    process::{Command, Stdio},
+};
 
+// Trait for all compression and decompression functions
 pub trait Compressor {
```
Comment (lhao03) [inline src-tauri/src/compressor.rs:10]: also can you edit main.rs to add the compression during encode_sequence? 

https://github.com/UBC-iGEM/dna-software/blob/151916379ad340278c7996866aa9d992d4827303/src-tauri/src/main.rs#L35-L47


it should follow the old deleted sequence.rs almost exactly

https://github.com/UBC-iGEM/dna-software/blob/2b88993ed61a17033ea4fb30c28932cfbc92ae6f/src-tauri/src/sequence.rs#L11-L21

**Review by lhao03 — CHANGES_REQUESTED** [review]

File: src-tauri/src/main.rs [hunk 3]
```diff
@@ -33,9 +33,16 @@ fn generate_primers(
...
+fn encode_sequence(
+    encoder_type: &str,
+    file_path: PathBuf,
+) -> Result<Vec<Base>, String> {
+    let compressor = VoidCompressor{}; 
+    let compressed = compressor
+        .compress(file_path)
+        .map_err(|err| err.to_string())?;
+    let bytes = fs::read(compressed).map_err(|err| err.to_string())?;
+    let bits = BitVec::<_, Msb0>::from_slice(&bytes);
```
Comment (lhao03) [inline src-tauri/src/main.rs:45]: To test this, do the below
image

**Review by lhao03 — CHANGES_REQUESTED** [review]

File: src-tauri/src/compressor.rs [hunk 4]
```diff
@@ -1,12 +1,182 @@
...
+
+    fn compress_lz4(&self, inpath: PathBuf) -> Result<PathBuf> {
+        dbg!("here");
+        let mut input_file = File::open(inpath.as_path())?;
+        dbg!("here");
+        let filename = inpath.file_stem().unwrap();
+        dbg!("here");
+        let out_dir = "compressed_lz4/";
+        fs::create_dir_all(out_dir);
+        let outpath = Path::new(out_dir).join(filename).with_extension("lz4");
```
Comment (lhao03) [inline src-tauri/src/compressor.rs:94]: The assumption that this directory exists causes an error to be thrown; additionally, there is no handling of the case where the directory does not exist; I pushed a way to fix this, but this assumption is made in several other places in the code; please fix them and add proper handling of the case when the directory doesn't exist (either make it through Rust, or prompt the user to make it themselves).

**Review by lhao03 — CHANGES_REQUESTED** [review]

**Comment by seb-hyland** [comment]
Hi Lucy, I pushed up a hefty merge with my changes, your changes, and changes in main. This should include:
1. Your changes to `compress_lz4` to ensure the directory exists
2. My updates to `compressor.rs` and the updated `encode_sequence` function with my changes
3. I have kept your `dbg!("here");` macros; do you wish to keep those for now?

Although I haven't gone over everything again since the merge (except for resolving conflicts), the project is compiling and tests are passing (including the quickcheck for (de)compression). Do you wish to make a test for `encode_sequence`? I'm not sure how we would do that save for a test similar to `chaosDNA` (no assertion, just use some test sentence and print the output to CLI).

**Comment by lhao03** [comment]
> Hi Lucy, I pushed up a hefty merge with my changes, your changes, and changes in main. This should include:
> 
> 1. Your changes to `compress_lz4` to ensure the directory exists
> 2. My updates to `compressor.rs` and the updated `encode_sequence` function with my changes
> 3. I have kept your `dbg!("here");` macros; do you wish to keep those for now?
> 
> Although I haven't gone over everything again since the merge (except for resolving conflicts), the project is compiling and tests are passing (including the quickcheck for (de)compression). Do you wish to make a test for `encode_sequence`? I'm not sure how we would do that save for a test similar to `chaosDNA` (no assertion, just use some test sentence and print the output to CLI).

Yeah we can have a test for encode_sequence using quickcheck, I can set that up (but yeah it would be a test sentence or str that we can check compress -> bit -> base -> bit -> decompress)

**Review by lhao03 — APPROVED** [review]
Thanks for working on this Sebastian!

**Pull request merged**